### ISSUE TITLE
[no-param-reassign] Ignore rest properties and arguments

### ISSUE THREAD

**Issue by gfpacheco** [eslint/eslint#9431]
**What rule do you want to change?**
no-param-reassign

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behaviour

**Please provide some example code that this change will affect:**

```js
function foo({ bar, ...rest }) {
  rest.baz = 'baz';
}

// and

function foo(bar, ...rest) {
  rest[0] = 'baz';
}
```

**What does the rule currently do for this code?**
Shows error: `Assignment to property of function parameter 'rest'.`

**What will the rule do after it's changed?**
Does not show error.

---

Since the variable is already a new object, it doesn't make sense the need to copy it again before altering it.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.

**Comment by gfpacheco** [comment]
Can I at least you tell me why you don't think this is a good idea? Maybe I change my opinion too, because for now I just don't see any reason why this would be bad

**Comment by not-an-aardvark** [comment]
I don't think it's necessarily a *bad* idea -- it's just that there wasn't consensus among the team to change the behavior. We get a lot of feature/change requests, and we can't realistically maintain all of them, so we only accept enhancement which would have a very clearly positive impact.
